### PR TITLE
Add more detail to the Lambda Annotations README.md

### PR DESCRIPTION
**Pull request by normj**
*Description of changes:*
Rework the Lambda Annotations README file

* Reordered the content to call specific major features
* Added a table of contents to the top of the page
* Added API Gateway example
* Added S3 Event example

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 1]
```diff
@@ -1,12 +1,26 @@
...
 
-The Lambda Annotations is a programming model for writing .NET Lambda function. At a high level the programming model allows
-idiomatic .NET coding patterns and uses [C# Source Generators](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview) to bridge the gap between the Lambda programming model 
-to the more idiomatic programming model.
-
-For example here is a simplistic example of a .NET Lambda function that acts like a calculator plus method using the normal 
-Lambda programming model. It respondes to an API Gateway REST API, pulls the operands from the resource paths, does the 
+The Lambda Annotations is a programming model for writing .NET Lambda functions. At a high level the programming model allows
+idiomatic .NET coding patterns and uses [C# Source Generators](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview) to bridge the 
+gap between the Lambda programming model to the more idiomatic programming model.
```
Comment (eddiemcs3) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:5]: Redundant use of idiomatic

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 2]
```diff
@@ -53,9 +68,113 @@ public class Functions
...
+    }
+}
+```
+
+## Synchronizing CloudFormation template
+
+When the .NET project is compiled the Lambda Annotation source generator will synchronize all of the C# methods with the `LambdaFunction` attribute in the 
+project's CloudFormation template. Support is available for both JSON and YAML based CloudFormation templates.
+The source generator identifies the CloudFormation template for the project by looking at the `template` property in the `aws-lambda-tools-defaults.json` 
+file. If either the property or file does not exist the source generator defaults to `serverless.template` and creates the file if it doesn't exist.
```
Comment (eddiemcs3) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:132]: Could we reword this to something like "If the 'template' property is absent, the source generator will default to the 'serverless.template' and create the file if it does not exist.

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 3]
```diff
@@ -68,37 +187,141 @@ public class Functions
...
+        }
+      }
+    },
+```
+
+## Amazon API Gateway example
+
+This example creates a REST API through Amazon API Gateway that exposes the common arithmetic operations. 
+
+To avoid putting business logic inside the REST API a separate calculator service is created to encapsulated the logic of the arithmetic operations. Here is both the
```
Comment (eddiemcs3) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:217]: To avoid putting business logic inside the REST API a separate calculator service is created to encapsulate the logic of the arithmetic operations. Here is both the 
calculator service's interface and default implementation. 
*Typo fix

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 4]
```diff
@@ -68,37 +187,141 @@ public class Functions
...
+        services.AddSingleton<ICalculatorService, DefaultCalculatorService>();
+    }
+}
 ```
 
-  "Parameters": {
-    "LambdaExecutionRole": {
-      "Type": "String"
+Since the `ICalculatorService` is registered as a singleton the service is injected into the Lambda function via the constructor. 
+If the registered service is registered as scoped or transient and a new instance is need for each Lambda invocation then the
```
Comment (eddiemcs3) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:260]: If the registered service is registered as scoped or transient, and a new instance is needed for each Lambda invocation, then the
`FromServices` attribute should be used on a method parameter of the Lambda function. 

*Typo fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 5]
```diff
@@ -68,37 +187,141 @@ public class Functions
...
 ```
 
-The above two examples when used together will the use the value of LambdaRoleParameter as the role during deployment.
+For each arithmetic operation a separate C# method is added containing the `LambdaFunction` attribute. The `LambdaFunction` attribute
+ensures the dependency injection framework is hooked up to the Lambda function and the Lambda function will be declared in the 
+CloudFormation template. 
 
-## Source Generator
+Since these Lambda functions are responding to API Gateway events the `HttpApi` attribute is added 
+to registered the event source in CloudFormation along with the HTTP verb and resource path. The `HttpApi` attribute also enables
```
Comment (eddiemcs3) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:280]: "to registered the event source in CloudFormation" should be "to register the event source..." ?

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 6]
```diff
@@ -68,37 +187,141 @@ public class Functions
...
+[HttpApi(LambdaHttpMethod.Get, "/divide/{x}/{y}")]
+public int Divide(int x, int y, ILambdaContext context)
+{
+    context.Logger.LogInformation($"{x} divide {y} is {x / y}");
+    return _calculatorService.Divide(x, y);
 }
 ```
-The source generator adds the following entry in the CloudFormation template corresponding to the above Lambda function:
+
+For each `LambdaFunction` declared the source generator will updated the CloudFormation template with the corresponding resource.
```
Comment (eddiemcs3) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:319]: For each `LambdaFunction` declared the source generator will update the CloudFormation template with the corresponding resource. 
*Typo fix

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 7]
```diff
@@ -68,37 +187,141 @@ public class Functions
...
+public int Divide(int x, int y, ILambdaContext context)
+{
+    context.Logger.LogInformation($"{x} divide {y} is {x / y}");
+    return _calculatorService.Divide(x, y);
 }
 ```
-The source generator adds the following entry in the CloudFormation template corresponding to the above Lambda function:
+
+For each `LambdaFunction` declared the source generator will updated the CloudFormation template with the corresponding resource. 
+Here you can see the Lambda CloudFormation resource has the `Handler` property to the generated code by the source generator.
```
Comment (eddiemcs3) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:320]: Awkward.  "Here you can see that the Handler property has been added to the generated code for the Lambda CloudFormation resource." ?

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 8]
```diff
@@ -111,84 +334,283 @@ The source generator adds the following entry in the CloudFormation template cor
...
 
-To see the code that is generated by the source generator turn the verbosity to detailed when executing a build. From the command this 
-is done by using the `--verbosity` switch.
-```
-dotnet build --verbosity detailed
-```
-To change the verbosity in Visual Studio go to Tools -> Options -> Projects and Solutions and adjust the MSBuild verbosity drop down boxes.
+Here is an example of the generated code from the source generator for the `Add` Lambda function. The generated code wraps around the 
+C# method that has the `LambdaFunction` attribute. It takes care of
+configuring the dependency injection, get the parameters from the API Gateway event and invoked the wrapped `LambdaFunction`. This code snippet is hear for
```
Comment (eddiemcs3) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:355]: "configuring the dependency injection, gets the parameters from the API Gateway event, and invokes the wrapped `LambdaFunction`.
*Typo

**Review by eddiemcs3 — COMMENTED** [review]
Added comments.

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 1 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:5]: Reworked

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 2 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:132]: Done

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 3 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:217]: Fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 4 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:260]: Fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 5 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:280]: Fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 6 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:319]: Fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 8 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:355]: Fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 7 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:320]: Reworked

**Comment by normj** [comment]
Responded to Ed's comments. Also went back through and removed usage of "second person" narrative to be consistent.

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 9]
```diff
@@ -224,8 +647,4 @@ parameter to the `LambdaFunction` must be the event object and the event source
 
 ## Project References
 
-If you are using API Gateway event attributes, such as `RestAPI` or `HttpAPI`, you need to manually add a package reference to `Amazon.Lambda.APIGatewayEvents` in your project, otherwise the project will not compile. We do not include it by default in order to keep the `Amazon.Lambda.Annotations` library lightweight.
-
-This release only supports API Gateway Events. As we add support for other types of events, such as S3 or DynamoDB, the list of required package references will depend on the Lambda .NET attributes you are using. 
-
-**Note**: If you are using [dependency injection](#dependency-injection) to write functions for other service events , such as S3 or DynamoDB, you will need to reference these packages in your project as well
```
Comment (normj) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:231]: I purposely deleted this `Note` because after rereading I realized it doesn't make sense. It was basically saying if you are using dependency injection in non API Gateway cases be sure to add your event package but those tasks are unrelated. 

It mattered in the API Gateway case, but not for DI, because the source generator is using the event package on behalf of the user. In the non API Gateway case the user is directly using the package.

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 10]
```diff
@@ -53,9 +68,113 @@ public class Functions
...
+    }
+}
+```
+
+## Synchronizing CloudFormation template
+
+When the .NET project is compiled the Lambda Annotation source generator will synchronize all of the C# methods with the `LambdaFunction` attribute in the 
+project's CloudFormation template. Support is available for both JSON and YAML based CloudFormation templates.
+The source generator identifies the CloudFormation template for the project by looking at the `template` property in the `aws-lambda-tools-defaults.json` 
+file. If the `template` property is absent, the source generator will default to `serverless.template' and create the file if it does not exist.
```
Comment (ashovlin) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:132]: nit: Wrong backtick if you were trying to highlight `serverless.template`

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 11]
```diff
@@ -68,37 +187,142 @@ public class Functions
...
+
+To avoid putting business logic inside the REST API a separate calculator service is created to encapsulate the logic of the arithmetic operations. Here is both the 
+calculator service's interface and default implementation.
+
+```csharp
+public interface ICalculatorService
+{
+    int Add(int x, int y);
+
+    int Substract(int x, int y);
```
Comment (ashovlin) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:225]: Typo in "Subtract" throughout both code blocks.

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 12]
```diff
@@ -1,12 +1,26 @@
 # Amazon.Lambda.Annotations
 
-The Lambda Annotations is a programming model for writing .NET Lambda function. At a high level the programming model allows
-idiomatic .NET coding patterns and uses [C# Source Generators](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview) to bridge the gap between the Lambda programming model 
-to the more idiomatic programming model.
-
-For example here is a simplistic example of a .NET Lambda function that acts like a calculator plus method using the normal 
-Lambda programming model. It respondes to an API Gateway REST API, pulls the operands from the resource paths, does the 
+The Lambda Annotations is a programming model for writing .NET Lambda functions. At a high level the programming model allows
```
Comment (ashovlin) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:3]: nit: agreement feels off here, I'd probably just drop the first "The"

Thought about "...annotations _are_...", but fine if it's a singular "model"

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 13]
```diff
@@ -53,9 +68,113 @@ public class Functions
...
+```
+
+## Synchronizing CloudFormation template
+
+When the .NET project is compiled the Lambda Annotation source generator will synchronize all of the C# methods with the `LambdaFunction` attribute in the 
+project's CloudFormation template. Support is available for both JSON and YAML based CloudFormation templates.
+The source generator identifies the CloudFormation template for the project by looking at the `template` property in the `aws-lambda-tools-defaults.json` 
+file. If the `template` property is absent, the source generator will default to `serverless.template' and create the file if it does not exist.
+
+The source generator synchronizes Lambda resources in the CloudFormation template. The template can still be edited to add additional AWS resources or further customize the Lambda functions to add
```
Comment (ashovlin) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:134]: nit: Second sentence on this line is tough to follow.

Maybe something like _"The template can still be edited to add additional AWS resources or to further customize the Lambda functions, such as adding other event sources that are not currently supported by the Lambda Annotations attributes."_

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 14]
```diff
@@ -111,84 +335,283 @@ The source generator adds the following entry in the CloudFormation template cor
...
 
-To see the code that is generated by the source generator turn the verbosity to detailed when executing a build. From the command this 
-is done by using the `--verbosity` switch.
-```
-dotnet build --verbosity detailed
-```
-To change the verbosity in Visual Studio go to Tools -> Options -> Projects and Solutions and adjust the MSBuild verbosity drop down boxes.
+Here is an example of the generated code from the source generator for the `Add` Lambda function. The generated code wraps around the 
+C# method that has the `LambdaFunction` attribute. It takes care of
+configuring the dependency injection, get the parameters from the API Gateway event and invokes the wrapped `LambdaFunction`. This code snippet is hear for
```
Comment (ashovlin) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:356]: nits:
* get -> gets
* hear -> here

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 15]
```diff
@@ -111,84 +335,283 @@ The source generator adds the following entry in the CloudFormation template cor
...
+          }
+        }
+      }
+    },
+```
+
+This is the code the source generator will produce for this function. The constructor is handling setting up the dependency injection. During the generated `Resize`
+method a dependency injection scope is created and then the `IImageServices` is retrieved from the dependency injection and passed into the function written 
+by the developer. By creating the scope in the generated `Resize` method all services registered as scoped or transient will trigger a new instance to be created
+when retrieved from the dependency injection framework. This code snippet is hear for
```
Comment (ashovlin) [inline Libraries/src/Amazon.Lambda.Annotations/README.md:567]: nit: hear -> here

**Review by ashovlin — CHANGES_REQUESTED** [review]

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 12 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:3]: Removed the `The`

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 16]
```diff
@@ -53,9 +68,113 @@ public class Functions
 }
 ```
 
-## Using References To Other Resources and Parameters in the template
+Lambda Annotations uses C# source generators to generate that boiler plate code to brige the gap between the default Lambda programming model to Lambda Annotations programming model at compile time.
```
[the remark this answers is not in the document]
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:71]: fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md [hunk 17]
```diff
@@ -53,9 +68,113 @@ public class Functions
...
+In addition the source generator also synchonizes the CloudFormation template to declare all of the .NET methods with the `LambdaFunction` attribute as
```
[the remark this answers is not in the document]
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:72]: fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 10 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:132]: Fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 13 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:134]: Took your suggestion

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 11 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:225]: Fixed. Something about that word always messes me up.

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 14 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:356]: Fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 15 above)
Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md:567]: fixed

File: Libraries/src/Amazon.Lambda.Annotations/README.md (hunk 11 above)
Comment (ashovlin) [reply Libraries/src/Amazon.Lambda.Annotations/README.md]: Still 4 more instances of `substract` in the next codeblock with the function implementations

Comment (normj) [reply Libraries/src/Amazon.Lambda.Annotations/README.md]: Fixed

**Review by ashovlin — APPROVED** [review]

**Review by 96malhar — APPROVED** [review]
